pgxn/neon/libpagestore.c

```diff
@@ -100,29 +113,44 @@ AssignPageserverConnstring(const char *newval, void *extra)
 {
 	if (!PagestoreShmemIsValid())
 		return;
-	LWLockAcquire(pagestore_shared->lock, LW_EXCLUSIVE);
+	pg_atomic_add_fetch_u64(&pagestore_shared->begin_update_counter, 1);
 	strlcpy(pagestore_shared->pageserver_connstring, newval, MAX_PAGESERVER_CONNSTRING_SIZE);
-	pg_atomic_fetch_add_u64(&pagestore_shared->update_counter, 1);
-	LWLockRelease(pagestore_shared->lock);
+	pg_atomic_add_fetch_u64(&pagestore_shared->end_update_counter, 1);
 }
```

Comment:
There's nothing here that guarantees that only postmaster tries to set this. Two backends can enter this function concurrently, and if they try to set different values, you can end up with a mix of both strings.

Reply:
> Alternative idea: Instead of insisting that the postmaster updates the value in shared memory, keep the current > lwlock, and have all other processes update it except postmaster?

May be you forgot, but the original problem (why do we need to place this GUC value in shared memory) was that config is not reloaded during query execution. And we really need to change connection string to PS to be able to complete query. Only postmaster is never executing any query and so its should always be able to apply GUC changes.

Reply:
> There's nothing here that guarantees that only postmaster tries to set this. Two backends can enter this function concurrently, and if they try to set different values, you can end up with a mix of both strings.

Sorry, I missed this check when extracting this changes in separate PR.
Fixed.

Reply:
> > There's nothing here that guarantees that only postmaster tries to set this. Two backends can enter this function concurrently, and if they try to set different values, you can end up with a mix of both strings.
> 
> Sorry, I missed this check when extracting this changes in separate PR. Fixed.

I don't see the fix

Reply:
Sorry, fix is pushed now.
As well as test.